### PR TITLE
Introduce taker ActorSystem

### PR DESCRIPTION
**Pull request by klochowicz**
The connection actor is a real actor yet, however it should be enough to
abstract away from the production I/O.

### CODE REVIEW

**Comment by klochowicz** [comment]
note: the best way forward for a connection actor seems to be allowing multiple makers (or at least operation without the maker, which makes sense - there's no reason for not allowing to monitor our ongoing cfd's after maker goes offline).

File: daemon/src/taker_cfd.rs
Comment (thomaseizinger) [inline daemon/src/taker_cfd.rs:95]: I am happy to merge minor fixes like this so if you want to do this, please make a separate commit / PR :)

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
Minor fixes like imports and stuff can still go in. I would like to hold off with refactorings like the `ActorSystem` until we did the release :)

Meaning we can either split out the minor changes as commits or we just pause the entire one until we did the release.

File: daemon/src/taker_cfd.rs
Comment (luckysori) [reply daemon/src/taker_cfd.rs:95]: This is already in master. This branched wasn't rebased yet.

Comment (klochowicz) [reply daemon/src/taker_cfd.rs:95]: @luckysori for some reason, this one didn't land in the master - I think only maker fixes did. I was fairly sure I rebased against master before posting the PR, and doublechecked now. 
However, at this stage I might as well hold off till after the release 😉

**Comment by klochowicz** [comment]
> Minor fixes like imports and stuff can still go in. I would like to hold off with refactorings like the `ActorSystem` until we did the release :)
> 
> Meaning we can either split out the minor changes as commits or we just pause the entire one until we did the release.

Happy to pause it, as the release is not far off :)

File: daemon/src/taker_cfd.rs
Comment (luckysori) [reply daemon/src/taker_cfd.rs:95]: I'm silly, I removed the `+ Send` from the field, but not from the argument :sweat_smile:

**Comment by klochowicz** [comment]
> Minor fixes like imports and stuff can still go in. I would like to hold off with refactorings like the `ActorSystem` until we did the release :)
> 
> Meaning we can either split out the minor changes as commits or we just pause the entire one until we did the release.

@thomaseizinger made a release earlier today - we can think about merging this one now 💭

File: daemon/src/taker.rs
Comment (thomaseizinger) [inline daemon/src/taker.rs]: `async move` for a single `.await` expression should be equivalent to just passing the initial future in!

```suggestion
monitor::Actor::new(electrum, channel, cfds.clone())
```

File: daemon/src/connection.rs
Comment (thomaseizinger) [inline daemon/src/connection.rs]: This will block the thread, better use `tokio`'s `async` sleep!

File: daemon/src/connection.rs
Comment (thomaseizinger) [inline daemon/src/connection.rs:12]: Any chance we can have this split out in its own commit easily?

If it is too hard, don't bother.

File: daemon/src/taker.rs
Comment (thomaseizinger) [inline daemon/src/taker.rs:181]: I think with https://github.com/comit-network/hermes/pull/381, there is actually no more reason why we cannot boot up the actor system outside of rocket.

Doing that would allow us for easier error handling because we can just use `?` instead of `unwrap`ing in here.

Best to be done in a follow-up though!

File: daemon/src/taker.rs
Comment (thomaseizinger) [inline daemon/src/taker.rs]: :stop_sign: Why are we passing the same actor twice here? That can't be right!

File: daemon/src/taker.rs
Comment (thomaseizinger) [inline daemon/src/taker.rs:278]: Is it worth it to do this inside here if we are passing the `cfd_feed_receiver` out again? We could just do this outside and only pass the `sender` in?

File: daemon/src/taker.rs
Comment (thomaseizinger) [inline daemon/src/taker.rs:279]: I think the same applies here. I'd rather pass the `order_feed_sender` in than passing the `order_feed_receiver` out.

File: daemon/src/taker.rs
Comment (thomaseizinger) [inline daemon/src/taker.rs:280]: If we do the same thing with all of these, we can reduce the return values to just the `cfd_actor::Address`!

File: daemon/src/taker.rs
Comment (thomaseizinger) [inline daemon/src/taker.rs:269]: For the tests, we might need control over the oracle and monitor actors so you will likely need to return their addresses (`O` and `F`) from this function!

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
Found a bug and noted some potential improvements :)

Thanks!

File: daemon/src/taker.rs
Comment (klochowicz) [reply daemon/src/taker.rs]: it was a bit weird for me too, but I copied it from already existing pattern:
https://github.com/comit-network/hermes/blob/taker-actor-system/daemon/src/maker.rs#L356

(shall I fix it both places? :) )

File: daemon/src/taker.rs
Comment (klochowicz) [reply daemon/src/taker.rs:269]: > Found a bug and noted some potential improvements :)
> 
> Thanks!

This ActorSystem closely matches the one on the maker side - perhaps it would be best for me to create a follow-up that rolls out the suggestions that apply to both sides? (I'll address the ones that solely matter for the taker here)

File: daemon/src/taker.rs
Comment (thomaseizinger) [reply daemon/src/taker.rs]: Yes!

File: daemon/src/taker.rs
Comment (klochowicz) [reply daemon/src/taker.rs:278]: well, it saves some of the boilerplate.... and I'm just repeating the pattern from the maker. 
I don't have any strong feelings for or against, but I see it as a bit more convenient.

File: daemon/src/taker.rs
Comment (klochowicz) [reply daemon/src/taker.rs]: done! I'll attach the maker commit to this PR as well.

File: daemon/src/taker.rs
Comment (klochowicz) [reply daemon/src/taker.rs:280]: As stated above, happy to do it, but it will result in more boilerplate in the tests, so we'll need to add one more wrapper that adds these watch channels (perhaps that's better, as it splits the responsibilities?)

File: daemon/src/taker.rs
Comment (klochowicz) [reply daemon/src/taker.rs:181]: happy to do it as a follow-up :)

File: daemon/src/connection.rs
Comment (klochowicz) [reply daemon/src/connection.rs:12]: not too easily, I'll keep it mind to be more atomic with my commits!

**Comment by klochowicz** [comment]
oops, I accidentally autosquashed, after `git absorb` resulted in 4 fixup commits :( sorry for making it more difficult to look at it again!

**Review by thomaseizinger — APPROVED** [review]

File: daemon/src/maker.rs
Comment (thomaseizinger) [inline daemon/src/maker.rs:352]: Pretty sure this bug has consequences in monitoring for CETs in the current release as the monitoring actor will not learn about attestations and hence not monitor for CETs.

cc @da-kami @bonomat

Comment (bonomat) [reply daemon/src/maker.rs:352]: good catch. I think @da-kami noticed this bug already.

Comment (thomaseizinger) [reply daemon/src/maker.rs:352]: We should push out a 0.1.1 that is branched off the 0.1.0 tag.

**Pull request merged**
